Singleton sync and chip loader

mattcarter11 opened the pull request:
**Feature / Changes**
- Include user uploaded songs, albums
- Make syncing of songs, albums, playlists, artist faster and only allow one sync to run (singleton like)
   - Before that the app crashed when loading +7K songs (#53)
- Show loading spinner in:
   - Chip when syncing the showed data
   - Liked playlist
-  Add a Downloads chip in all old view. This shows albums, artists, playlist that have one or more downloaded song
   - songs now have a dateDownload
   - Fetching of download entities with downloads, ... is purely SQL now = better performance
- Show N download/M songs in playlists/albums/artists
- migrate SongByArtistAsc logic to SQL to improve performance
- Break up DatabaseDao into smaller interfaces to improve development
- Concentrated some SQL selects into one function to centralize code and remove duplicate logic

<p><strong>Sync spinner in chips and liked playlist</strong></p>
<img src="https://github.com/user-attachments/assets/e86b8aa4-027b-42bd-88b8-ac13ce945a5b" width="200"/>
<img src="https://github.com/user-attachments/assets/e864c090-009f-45d1-b3df-e08820d9c740" width="200"/>
<img src="https://github.com/user-attachments/assets/4ca9b2fc-05c8-4d5f-a887-55ba13e06f4b" width="200"/>

<p><strong>New Download chip (show only playlists/arts with any downloaded song)</strong></p>
<img src="https://github.com/user-attachments/assets/054958b9-5d95-4199-89c0-ec1c0925ccd6" width="200"/>

<p><strong>N/M songs</strong></p>
<img src="https://github.com/user-attachments/assets/5ccd82e6-77d1-46d8-863c-89a28f78bf24" width="200"/>
<img src="https://github.com/user-attachments/assets/ad6633bd-b093-4b6c-b0b4-41c4ea1fbb5a" width="200"/>
<img src="https://github.com/user-attachments/assets/3c539166-12e6-49f8-a789-01b537c8f970" width="200"/>
<img src="https://github.com/user-attachments/assets/dfa4f0aa-9788-4439-88dd-2be986b93efe" width="200"/>
<img src="https://github.com/user-attachments/assets/ad53deb6-583c-4d7d-ba8c-7ba84c5c727a" width="200"/>
<img src="https://github.com/user-attachments/assets/e8913194-60fb-4541-b661-1061165597a6" width="200"/>
<img src="https://github.com/user-attachments/assets/64331c81-d4f6-45d5-8e17-d3f34c252eb3" width="200"/>
<img src="https://github.com/user-attachments/assets/163edb0c-af46-4f5f-a2d9-66736014c031" width="200"/>

**Fixes**
- Fix sort ByPlayCountAsc using count instead of sum

mattcarter11 committed:
- Include uploaded songs
- Make sync of songs, albums, playlists, arits faster (corutines) and only run as "singletons"
- Rename some functions

mattcarter11 committed:
refactor Selector Header and match sort header size

mattcarter11 committed:
Add loading in chips tokens when syncing that token

- Add loading in chips tokens when syncing that token
- Add library and download chips in playlists

mattcarter11 committed:
- add loading spinner in liked playlist
- only call syncs when ytmSync true

mattcarter11 committed:
- refactor to centralize the sort downloaded song logic

mattcarter11 committed:
fix OrderByPlaycount

mattcarter11 committed:
fix only sync when it should: when ytmSync disabled or the token is correct

mattcarter11 committed:
finish playlist download chip functionality

mattcarter11 committed:
- Add Download chip to artists and albums - Break up DatabaseDao into smaller interfaces

mattcarter11 committed:
fix SelectHeader not using dev changes on rebase

mattcarter11 committed:
show sync of playlist only in liked songs

mattcarter11 committed:
fix app crashing when no connection (use onSuccess instead of getOrThrow)

mattcarter11 commented:
@mikooomich this one has some big files 😅 if anything has to change, I'm more than happy to oblige

mikooomich commented:
As in a "this is ready for merge" or suggestions in general? I haven't looked into it too deeply, but looks great so far! A few things I noticed
- App crashes on database migration. You probably can just let auto migration handle it (you basically don't have to do anything for simple field additions), and onPostUpgrade() if necessary. The manual migrations are only really necessary for when auto migrations won't cut it
- On old layout (maybe new too?), switching tabs causes sync(?) even when auto sync is disabled

mattcarter11 commented:
@mikooomich db migration removed and fixed the ytmSync being ignored in the playlist tab

mattcarter11 commented:
@mikooomich also, what do you think about adding this? A n/m download count on playlists/artists/albums with any downloaded song.

Maybe add an option to toggle this?

<img src="https://github.com/user-attachments/assets/d8fe6738-2c82-4736-9244-e6d888510243" width="200"/>

mikooomich commented:
You cannot modify old db schema like that. You still need to increment db version and specify a database auto migration, see here -> https://github.com/DD3Boh/OuterTune/commit/c3e2ea7dc259209b10ab6666093a0a9f3f9c92c5#diff-841cb2099b74da4d913f1cf1d60c554d9e6ae0dc6f8c3a22ad3b95a4b5b39568R107

> A n/m download count on playlists/artists/albums with any downloaded song. Maybe add an option to toggle this?

Yeah, nice touch



If the pull request is wip could you mark it as such (with the selector), then mark it ready when its ready to be reviewed/merged

mattcarter11 committed:
add n/M download count to albums/artists/playlists

- add n/M download count to albums/artists/playlists
- move entity and maps queries to separate Daos

mattcarter11 commented:
@mikooomich change done and added N/M songs

mattcarter11 committed:
sync uploaded albums and artists

mikooomich commented on app/schemas/com.dd3boh.outertune.db.InternalDatabase/15.json:
This file should not be changed at all.  If you build DB 15 (before your DB16 changes), it should not modify this file from what is currently on dev branch

though, from the looks of it, you just forgot to revert this file's changes

mikooomich commented on app/src/main/java/com/dd3boh/outertune/extensions/ContextExt.kt:
```diff
@@ -0,0 +1,18 @@
+fun Context.isSyncEnabled(): Boolean {
+    return runBlocking {
+        val ytmSync = dataStore.data.map { it[YtmSyncKey] ?: true }.first()
+        val cookie = dataStore.data.map { it[InnerTubeCookieKey] ?: "" }.first()
```
why not dataStore.get()

mikooomich commented on app/src/main/java/com/dd3boh/outertune/db/MusicDatabase.kt:
```diff
@@ -313,6 +314,12 @@ val MIGRATION_14_15 = object : Migration(14, 15) {
     }
 }
 
+val MIGRATION_15_16 = object : Migration(15, 16) {
```
Auto migration does fail. Odd.

mikooomich commented on app/src/main/java/com/dd3boh/outertune/db/daos/PlaylistsDao.kt:
(Referring to all daos, not just playlist) Is there a pattern to what goes in the the specific daos vs the main dao?

For example `fun update(map: PlaylistSongMap)` is in the main dao, but `fun update(playlist: PlaylistEntity)` is in its own dao. Is this because the stuff that goes into the specific daos are entity specific? And then the main dao handles mapping and stuff?

But then shouldn't this be in the main dao?
```
   @Query("DELETE FROM playlist_song_map WHERE playlistId = :playlistId")
    fun clearPlaylist(playlistId: String)
```

Of course, this refactor is miles better than the monolithic DatabaseDao mess, but plz define what goes into where (maybe a comment block at the top of DatabaseDao.kt?), and be consistent

mikooomich reviewed (changes requested):
Hello. A few nitpicks, otherwise, looks great.

For commits, mostly for sakes of readability/traceability: 
- Please remove merge commit(s). PRs should be rebased on conflict, or left as-is 
- Remove any rebase fixup commits (those should be applied within commits themselves)
- Squash any duplicates ex. "fix sort lined song"
- All database schema should be contained within one commit. It's just the one change with SongEntity, right? All the json modifications, adding migration, etc can probably go into one commit
- Tagging.  You can leave this PR as is, however in the future, we'd very much prefer commits to be tagged according to functionality/category. ex "ui: ..." "syncutils: ...", etc. It just makes searching/skimming/grouping much easier 





Download icon is misaligned 
![image](https://github.com/user-attachments/assets/5b25c80d-481d-40cd-9702-452e48011a6f)


NPE crash when clicking on albums

```
 FATAL EXCEPTION: main
                                                                                                    Process: com.dd3boh.outertunepr63, PID: 4785
                                                                                                    java.lang.NullPointerException: Parameter specified as non-null is null: method androidx.room.RoomSQLiteQuery.bindString, parameter value
                                                                                                    	at androidx.room.RoomSQLiteQuery.bindString(Unknown Source:2)
                                                                                                    	at com.dd3boh.outertune.db.DatabaseDao_Impl.__fetchRelationshipartistAscomDd3bohOutertuneDbEntitiesArtistEntity_1(DatabaseDao_Impl.java:12905)
                                                                                                    	at com.dd3boh.outertune.db.DatabaseDao_Impl.-$$Nest$m__fetchRelationshipartistAscomDd3bohOutertuneDbEntitiesArtistEntity_1(Unknown Source:0)
                                                                                                    	at com.dd3boh.outertune.db.DatabaseDao_Impl$97.call(DatabaseDao_Impl.java:7750)
                                                                                                    	at com.dd3boh.outertune.db.DatabaseDao_Impl$97.call(DatabaseDao_Impl.java:7717)
                                                                                                    	at androidx.room.CoroutinesRoom$Companion$createFlow$1$1$1.invokeSuspend(CoroutinesRoom.kt:129)
                                                                                                    	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
                                                                                                    	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:104)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
                                                                                                    	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
                                                                                                    	at java.lang.Thread.run(Thread.java:1012)
                                                                                                    	Suppressed: kotlinx.coroutines.internal.DiagnosticCoroutineContextException: [StandaloneCoroutine{Cancelling}@b352779, Dispatchers.Main.immediate]
2024-10-20 13:01:55.637  4785-4785  Process                 com.dd3boh.outertunepr63             I  Sending signal. PID: 4785 SIG: 9
```

mattcarter11 commented:
@mikooomich 
- resolved all comments
- merge removed
- Squash some commits
- alignment fixed
- NPE crash when clicking on albums -> fixed (missing group by album.id)

- All database schema should be contained within one commit -> should i squash commits? what if they are mixed with other stuff?
- Remove any rebase fixup commits (those should be applied within commits themselves) -> what do you mean applied within commits themselves? the commits from the branch used to rebas (dev)?

mattcarter11 committed:
DAOs: unify albums/playlist/artists gets

mikooomich commented:
> All database schema should be contained within one commit -> should i squash commits? what if they are mixed with other stuff?

"migrate get download logic to SQL (+performance) and add isDownload to song table"
For this, i'd seperate the rest of the sql stuff (another commit) and add is downloaded (one commit) . When you add the isdownloaded field, it comes it side effects: you need to handle migration and room will generate the json for db V16. that should all be contained in that commit. 

This also means squash the following since into that commit since that will be handled in the commit you add isdownloaded
- remove db migration 
- add new versiondb file
- re-add migration since auto was failing
- revert 15.json changes 
- and any more i'm missing (?)

> Remove any rebase fixup commits (those should be applied within commits themselves) -> what do you mean applied within commits themselves? the commits from the branch used to rebas (dev)?

"add missing import after rebase" should not be its own commit (unless it's alot of fixing and makes sense to otherwise), instead, add it into the commit that needs it. (if you don't know where it came from, honestly you can just leave it as it)

mattcarter11 committed:
Albums: sync both yt and uploaded

mattcarter11 committed:
SongEntity: add isDownload to song table and set value

mattcarter11 committed:
Dao: migrate get download logic to SQL (+performance)

mattcarter11 commented:
@mikooomich changes done
